python/mlx/utils.py

```diff
@@ -191,3 +191,45 @@ def tree_unflatten(tree):
         return l
     else:
         return {k: tree_unflatten(v) for k, v in children.items()}
+
+
+def tree_reduce(fn, tree, initializer=None, is_leaf=None):
+    """
+    Recursively applies a reducer function `fn` to the leaves of the Python tree `tree` and aggregates them into a single value.
+
+    This function simplifies complex tree structures into a single accumulated result by applying a provided function `fn` from the leaves upward. It is similar to the built-in `reduce` function but works over nested, tree-like data structures.
+
+    Example:
+        >>> from mlx.utils import tree_reduce
+        >>> tree = {"a": [1, 2, 3], "b": [4, 5]}
+        >>> tree_reduce(lambda acc, x: acc + x, tree, 0)
+        15
+
+    Args:
+        fn (callable): The reducer function that takes two arguments (accumulator, current_value) and returns the updated accumulator.
+        tree (Any): The Python tree to reduce. It can be any nested combination of lists, tuples, or dictionaries.
+        initializer (Any, optional): The initial value to start the reduction. If not provided, the first leaf value is used.
+        is_leaf (callable, optional): A function to determine if an object is a leaf, returning True for leaf nodes and False otherwise.
```

Comment:
```suggestion
        is_leaf (callable, optional): A function to determine if an object is a leaf, returning ``True`` for leaf nodes and ``False`` otherwise.
```